src/main/java/com/suse/salt/netapi/utils/SaltErrorUtils.java

```diff
@@ -0,0 +1,41 @@
+package com.suse.salt.netapi.utils;
+
+import com.suse.salt.netapi.errors.FunctionNotAvailable;
+import com.suse.salt.netapi.errors.ModuleNotSupported;
+import com.suse.salt.netapi.errors.SaltError;
+
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Utils for deriving {@link SaltError} based on the salt text output.
+ */
+public class SaltErrorUtils {
+
+    private static final Pattern FN_UNAVAILABLE =
+            Pattern.compile("'([^']+)' is not available.");
+    private static final Pattern MODULE_NOT_SUPPORTED =
+            Pattern.compile("'([^']+)' __virtual__ returned False");
+
+    /**
+     * Based on the salt text output, derive particular {@link SaltError}.
+     * @param saltOutput salt output
+     * @return salt error corresponding to given string
+     */
+    public static Optional<SaltError> deriveError(Optional<String> saltOutput) {
```

Comment:
Nice picture, but it'd be faster for all if you provided a constructive feedback (code snippet would suffice).